### ISSUE TITLE
add boolean preprocessor symbol GRIMOIRE or similar

### ISSUE THREAD

**Issue by elenzil** [jshrake/grimoire#61]
currently shaders can be copy-and-pasted from the grimoire environment right into ShaderToy (yay!) with the minor modification of removing the statements defining fragColor and main() which appear at the bottom. eg [the 'new' example](https://github.com/jshrake/grimoire/blob/master/examples/shadertoy-new/image.glsl).

however, if Grimoire injected something like `#define GRIMOIRE` before sending the shader code off for compilation, then that minor step of removing those lines could be eliminated.
ie the code would become

```
#ifdef GRIMOIRE
out vec4 fragColor;
void main() { mainImage(fragColor, gl_FragCoord.xy); }
#endif 
```

i'm note sure what the actual best symbol would be, eg if there's already precedent for something like ENVIRONMENT_GRIMOIRE etc, nor the best way of injecting such a symbol. if i were doing it, i'd probably just literally prepend the line `#define GRIMOIRE` to the front of each shader file, but that seems hackish.

an alternative approach might be to implement something similar to Shadertoy's *common tab*, but only functional in grimoire. eg grim_common.glsl.   then one could add whatever #defines one pleased in there.  however with a view to establishing a consistent way stuff is done, i'd probably opt for the other approach first.

**Comment by jshrake** [comment]
This is a good feature request!

>  if i were doing it, i'd probably just literally prepend the line #define GRIMOIRE to the front of each shader file, but that seems hackish.

I think that's fine as it produces something fairly simple that's easy to document. I like it.

>  an alternative approach might be to implement something similar to Shadertoy's common tab, but only functional in grimoire. eg grim_common.glsl. 

You can currently `#include "any_file_you_want.glsl"`, but you would need to take into account how this would impact your ability to port the code over to shadertoy :)

> then one could add whatever #defines one pleased in there. however with a view to establishing a consistent way stuff is done, i'd probably opt for the other approach first.

One feature I've considered is allowing users to specify a set of preprocessor symbols in the pass configuration. This could be useful for re-using the same shader in different ways across multiple passes.

**Comment by elenzil** [comment]
> I think that's fine 
cool!

> #include "any_file.glsl"
sweet!  ..  but causes an "unknown error" on shadertoy,
even bracketed by `#if 0`.  [example](https://www.shadertoy.com/view/ttBGz1)

> allowing users to specify a set of preprocessor symbols in the pass configuration
yeah, that could definitely be useful.
i think i'd still consider forcing something for grimoire in there,
just to get folks on the same page w/r/t the symbol, versus making up our own "using_grimoire" "env_grimoire", etc.

**Comment by jshrake** [comment]
> sweet! .. but causes an "unknown error" on shadertoy, even bracketed by #if 0.

`#include <any_file.glsl>` also works in grimoire and also works when bracketed by #if 0 in shadertoy.

**Comment by elenzil** [comment]
so it does, great!

**Comment by jshrake** [comment]
Closed by #62. @elenzil, you should now be able to use `#ifdef GRIMOIRE` in your shader code.

> One feature I've considered is allowing users to specify a set of preprocessor symbols in the pass configuration. This could be useful for reusing the same shader in different ways across multiple passes.

I'm tracking this in #64

**Comment by elenzil** [comment]
woohoo, this is great!
